rfcs/0065-Migrate-queue-to-postgres.md

```diff
@@ -4,36 +4,98 @@
-   * heroku postgres
-   * google cloud SQL postgres
-   * self-hosted postgres
-   * self-hosted cockroachdb (we are close to key-value usage)
- * We can host in the same data center as our web nodes (reducing latency)
-
-Cons:
-  * It's not a hand-off auto-scaling solution
-  * We can do expensive queries and bring the system down
+* A shared base of indexed data can allow services to make smarter decisions and
+  save money on provisioning and data storage
+* Azure is unreliable with frequent failures and inscrutible rate-limiting. While we are resilient
+to these errors, they do cause inefficiency and sometimes cascading failures
+* Storing data in postgres will enable features users desire:
+  * Queue introspection
+    * What tasks are pending right now?
+    * How many tasks are [actually pending and not cancelled](https://bugzilla.mozilla.org/show_bug.cgi?id=1434851)?
+  * Integration of task-related and worker-related status
+    * What tasks has this worker performed?
+  * Ad-hoc analytics
+    * How much compute time do we use per push?
+    * What do we pay for wpt?
+  * More sophisticated dependency resolution [in the case of task failure](https://bugzilla.mozilla.org/show_bug.cgi?id=1443503#c8)
+  * Dynamically changing task levels
 
 # Details
 
-There is an initial attempt at outlining what the database schema would look at here:
-https://public.etherpad-mozilla.org/p/jonasfj-queue-with-postgres
+## Transition
+
+The transition to Postgres will be done in two steps:
+1. Transfer the data from Azure to Postgres without data loss. This will not aim
+   at normalizing the data but rather prioritize a safe and easy transition.  In
+   other words, the data will be backed by Postgres, but still behaving like
+   Azure.
+2. Go fully into normalized SQL tables.
+
+Upon completion of step 1, deployers of Taskcluster will need to provide a DB
+URL (maybe several) then run a script to "load" data from Azure to Postgres
+during a downtime.
+
+There will be a version of Taskcluster that supports Postgres and not Azure,
+with the previous version supporting Azure and not Postgres. So upgrading to
+that version is something that happens in a downtime.
+
+## Schema Migration
+
+Taskcluster will write its own tooling to handle schema migrations and will
+have all migrations tested with sample data to ensure we get the
+expected schema. In addition, there will be one schema for the whole monorepo,
+so that schema updates happen atomically across all services.
+
+## Backups and Restores
+
+Teams operating Taskcluster will rely on the cloud provider's backup system to
+handle backups and restores.
+
+## DB Version Upgrades/Downgrades Compatibility
+
+When deploying a new version of Taskcluster, any necessary database
+upgrades/downgrades *must* be applied before any services are upgraded.
+Taskcluster will maintain the invariant that services expecting databse version
+S can interoperate with a database at version V as long as V >= S.
```

Comment:
This would allow us to roll back deployed code (meaning JS code) without rolling back the database.  I wonder if we need to support DB rollbacks?

I suppose supporting them is not that difficult, and may be handy in some circumstances (such as if the DB upgrade itself causes problems).

Reply:
> I wonder if we need to support DB rollbacks?

Yep. The section title and body has "/downgrades" appended to every mention of "upgrades"to make it clear that we're going to implement both DB upgrades and downgrades (rollbacks).

Reply:
Yep, it looks good.  I was just wondering if maybe we didn't need to support DB downgrades.  If they turn out to be hard, we can re-visit that consideration.

Reply:
Ah I see. In that case, I'm happy to remove it from the RFC.

Reply:
Done.